packages/flutter_reactive_ble/lib/src/reactive_ble.dart

```diff
@@ -290,54 +294,258 @@ class FlutterReactiveBle {
 
+  /// Performs service discovery on the peripheral.
+  ///
+  /// When discovery fails this method throws an [Exception].
+  ///
+  /// Use [getDiscoveredServices] to get the discovered services
+  Future<void> discoverAllServices(String deviceId) => _connectedDeviceOperator.discoverServices(deviceId);
+
+  /// Services can be discovered by:
+  /// - specify `servicesWithCharacteristicsToDiscover` in [connectToDevice] or [connectToAdvertisingDevice]. Or,
+  /// - by calling [discoverAllServices]
+  ///
+  /// Discovered services and charactersitcs remain valid for as long as the device stays connected. After reconnecting
+  /// to a device, [getDiscoveredServices] must be called again to get updated services and characteristics.
+  ///
+  /// Note: On Android, this method performs service discovery if one was done yet after connecting the device.
+  Future<List<Service>> getDiscoveredServices(String deviceId) async {
+    final discoveredServices = await _connectedDeviceOperator.getDiscoverServices(deviceId);
+
+    final services = _services[deviceId] ?? [];
+
+    for (final discoveredService in discoveredServices) {
+      final service = services.firstWhere(
+        (service) => _isMatchingService(service, discoveredService),
+        orElse: () {
+          final newService = Service._(
+            id: discoveredService.serviceId,
+            instanceId: discoveredService.serviceInstanceId,
+            deviceId: deviceId,
+          );
+          services.add(newService);
+          return newService;
+        },
+      );
+
+      for (final discoveredCharacteristic in discoveredService.characteristics) {
+        if (!service._characteristics.any((char) => _isMatchingCharacteristic(char, discoveredCharacteristic))) {
+          service._characteristics.add(Characteristic._(
+            id: discoveredCharacteristic.characteristicId,
+            instanceId: discoveredCharacteristic.characteristicInstanceId,
+            service: service,
+            lib: this,
+            isReadable: discoveredCharacteristic.isReadable,
+            isWritableWithoutResponse: discoveredCharacteristic.isWritableWithoutResponse,
+            isWritableWithResponse: discoveredCharacteristic.isWritableWithResponse,
+            isNotifiable: discoveredCharacteristic.isNotifiable,
+            isIndicatable: discoveredCharacteristic.isIndicatable,
+          ));
+        }
+      }
+      for (final char in service._characteristics.toList()) {
+        if (!discoveredService.characteristics
+            .any((discoveredCharacteristic) => _isMatchingCharacteristic(char, discoveredCharacteristic))) {
+          service._characteristics.remove(char);
+        }
+      }
+    }
+    for (final service in services.toList()) {
+      if (!discoveredServices.any((discoveredService) => _isMatchingService(service, discoveredService))) {
+        services.remove(service);
+      }
+    }
+
+    return services.toList();
```

Comment:
It is the same instance as `_services` and I don't want to give clients a mutable private field. (A `ListView` could be a way to avoid the copy but that seems a bit uncommon and the list isn't big)